### PR TITLE
Allow creation of Hive tables with non-existent HDFS paths

### PR DESCRIPTION
**Pull request by amitds1997**
Currently, creation of a table using `CREATE TABLE` on a non-existent
HDFS path throws an exception saying that the external location must
be a directory, which is misleading. This is due to a missing check
whether the path exists or not.

There are two approaches to solve this:
1. Throw exception that the path does not exist.
2. Automatically create non-existent path.

2nd one is the approach taken by Hive and Spark. Also, Trino does the
same when running the `CTAS` command. So, 2 seems like the way to go.

`org.apache.hadoop.fs.FileSystem#isDirectory` is deprecated, so we
switch to using `org.apache.hadoop.fs.FileStatus#isDirectory`.

Was tested locally on HDP 3.3. Do let me know if we need to write additional tests for this.

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java [hunk 1]
```diff
@@ -1078,7 +1078,8 @@ private void checkExternalPath(HdfsContext context, Path path)
     {
         try {
             if (!isS3FileSystem(context, hdfsEnvironment, path)) {
-                if (!hdfsEnvironment.getFileSystem(context, path).isDirectory(path)) {
+                FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, path);
+                if (fileSystem.exists(path) && !fileSystem.getFileStatus(path).isDirectory()) {
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1082]: nit: change the error message to: `"External location exists, but not a directory: " + path`

**Review by phd3 — CHANGES_REQUESTED** [review]
the direction seems good to me as it's consistent with CTAS, one difference though is that now CT would allow (both) with or without directory existence. cc @findepi if there was some  concern/discussion for intentional omission while implementing #2669.

@amitds1997 I think adding tests in `TestHiveCreateExternalTable` would be useful.

**Comment by amitds1997** [comment]
One point that needs mention here is that `CT` with an external location in S3, currently creates the location automatically, if it does not exist, without throwing any error.

**Comment by amitds1997** [comment]
Most of our tests are using `FileHiveMetastore` which does not create directories, most metastore implementations do (or am I wrong here?).
https://github.com/trinodb/trino/blob/f652531e06a1ec5afe86db0f8acf1a1f5274099f/plugin/trino-hive/src/main/java/io/trino/plugin/hive/metastore/file/FileHiveMetastore.java#L302-L313

Because of this, for writing passing tests, I would have to change the implementation of `FileHiveMetastore` to allow it to create directories. Am I missing something here or is this change indeed required?

@phd3

**Comment by phd3** [comment]
@amitds1997 I think we can modify the check, since it doesn't reflect what is being implemented. It'd be also useful to add a product test for this so we test against production HMS as well. (some examples in `io.trino.tests.hive.TestExternalHiveTable`)

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java (hunk 1 above)
Comment (amitds1997) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java:1081]: @phd3 Could you review this again please?

**Review by phd3 — APPROVED** [review]
Looks good to me. @findepi does this direction seem reasonable to you?

**Comment by findepi** [comment]
> does this direction seem reasonable to you?

it does

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java [hunk 2]
```diff
@@ -149,6 +149,38 @@ public void testCreateExternalTableWithInaccessibleSchemaLocation()
         String table = "test_create_external";
         String tableLocation = "/tmp/" + table;
         query(format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS SELECT * FROM tpch.tiny.nation", HIVE_CATALOG_WITH_EXTERNAL_WRITES, schema, table, tableLocation));
+
+        String otherSchema = "schema_with_unknown_location";
+        query(format("CREATE SCHEMA %s.%s", HIVE_CATALOG_WITH_EXTERNAL_WRITES, otherSchema));
+
+        String otherTable = "table_with_unknown_schema_location";
+        String otherTableLocation = "/tmp/" + otherTable;
+        query(format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS SELECT * FROM tpch.tiny.nation", HIVE_CATALOG_WITH_EXTERNAL_WRITES, otherSchema, otherTable, otherTableLocation));
```
Comment (phd3) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:158]: here, looks like the schema directory would still exist. can we just create this table in `schema` used for previous testcase ?

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java [hunk 3]
```diff
@@ -149,6 +149,38 @@ public void testCreateExternalTableWithInaccessibleSchemaLocation()
...
+    @Test(groups = {HIVE_WITH_EXTERNAL_WRITES, PROFILE_SPECIFIC_TESTS})
+    public void testCreateExternalTableWithExistingPath()
+    {
+        String schema = "schema_with_file_as_location";
+
+        String table = "table_with_file_as_location";
+        String tableLocation = "/tmp/" + table;
+        hdfsClient.saveFile(tableLocation, "This is a file");
+        assertThat(() -> query(
+                format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS SELECT * FROM tpch.tiny.nation", HIVE_CATALOG_WITH_EXTERNAL_WRITES, schema, table, tableLocation)))
```
Comment (phd3) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:170]: I think we'd want to use CT, and not CTAS here right? the error message should say that the path exists but not a directory.

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java (hunk 2 above)
Comment (phd3) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:158]: use CT.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveLocationService.java [hunk 4]
```diff
@@ -58,9 +58,9 @@ public LocationHandle forNewTable(SemiTransactionalHiveMetastore metastore, Conn
         HdfsContext context = new HdfsContext(session);
         Path targetPath = externalLocation.orElseGet(() -> getTableDefaultLocation(context, metastore, hdfsEnvironment, schemaName, tableName));
 
-        // verify the target directory for the table
+        // verify the target path for the table
```
Comment (phd3) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveLocationService.java:61]: can you please pull this out in a separate commit?

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java [hunk 5]
```diff
@@ -149,6 +149,38 @@ public void testCreateExternalTableWithInaccessibleSchemaLocation()
...
+
+        String otherTable = "table_with_unknown_schema_location";
+        String otherTableLocation = "/tmp/" + otherTable;
+        query(format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS SELECT * FROM tpch.tiny.nation", HIVE_CATALOG_WITH_EXTERNAL_WRITES, otherSchema, otherTable, otherTableLocation));
+    }
+
+    @Test(groups = {HIVE_WITH_EXTERNAL_WRITES, PROFILE_SPECIFIC_TESTS})
+    public void testCreateExternalTableWithExistingPath()
+    {
+        String schema = "schema_with_file_as_location";
```
Comment (phd3) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:164]: let's create the schema here, because we're not testing with non-existent schema. without schema it should also fail at some point during execution, but that's not the scope here.

**Review by phd3 — CHANGES_REQUESTED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java (hunk 2 above)
Comment (hashhar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:158]: Correct. IIUC we only need to perform a CREATE TABLE (not CTAS) here using the `schema_without_location`  schema.

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java (hunk 3 above)
Comment (hashhar) [reply testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:170]: I think we should have both here. A CT and a CTAS to verify that the CTAS behaviour matches the CT behaviour now.

File: testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java [hunk 6]
```diff
@@ -149,6 +149,38 @@ public void testCreateExternalTableWithInaccessibleSchemaLocation()
...
+        String tableLocation = "/tmp/" + table;
+        hdfsClient.saveFile(tableLocation, "This is a file");
+        assertThat(() -> query(
+                format("CREATE TABLE %s.%s.%s WITH (external_location = '%s') AS SELECT * FROM tpch.tiny.nation", HIVE_CATALOG_WITH_EXTERNAL_WRITES, schema, table, tableLocation)))
+                .failsWithMessage(format("Target path for table '%s.%s' already exists: %s", schema, table, tableLocation));
+        hdfsClient.delete(tableLocation);
+    }
+
+    @Test(groups = {HIVE_WITH_EXTERNAL_WRITES, PROFILE_SPECIFIC_TESTS})
+    public void testCreateExternalTableWithNonExistentPath()
```
Comment (hashhar) [inline testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestExternalHiveTable.java:176]: How is this different from the new case added in `testCreateExternalTableWithInaccessibleSchemaLocation`?

**Review by hashhar — DISMISSED** [review]
Looks almost good to go % some test changes.

Thanks for working on this @amitds1997. This should improve the out-of-box experience of CREATE TABLE because today you need to manually create directories on HDFS before you can CREATE TABLE there - which users shouldn't need to do.
